### ISSUE TITLE
Document alternative Clojure CLI usage

### ISSUE THREAD

**Issue by vemv** [clojure-emacs/enrich-classpath#48]
Besides from the suggested makefile, one could use clojure.sh directly (at the cost of losing some caching).

This is worth documenting, maybe with a suggested accompanying bash/zsh helper to use it concisely.

**Comment by anonimitoraf** [comment]
Yes please! I use the clojure CLI and would be keen to try this feature out

**Comment by vemv** [comment]
Thanks for the interest!

This repo bundles [clojure.sh](https://github.com/clojure-emacs/enrich-classpath/blob/92bc201fd19bb4adbcec73637fca0c685a6844e3/tools.deps/src/cider/enrich_classpath/clojure.sh).

If you use CIDER, it's also always bundled with each release and can be found in e.g. `~/.emacs.d./elpa/cider-*/clojure.sh` (or elsewhere if you use Git for managing packages).

So you can download it, or reference it (the latter is recommended for always having the latest version)

Now, asuming `clojure.sh` is in your PATH, you can, 

    clojure.sh clojure foo bar baz # pass any argument you'd normally pass to `clojure`

Where `clojure` is the name of your usual `clojure` command (it's `clojure` for 99% of users, but maybe you have multiple versions around or some other custom setup)

The usage is simple: it has the same arguments as `clojure`.

What the script does is:

* dynamically creating a new `clojure` invocation that will use a `-Scp` calculated by enrich-classpath; and then
* `eval`uate that invocation.

You can create a utility function that wraps the pattern above, adding it to your bash/zsh init file:

```bash
enrich(){
  ~/.emacs.d/elpa/cider*/clojure.sh clojure "$@"
}

# Usage:
# enrich foo bar baz # pass any argument you'd normally pass to `clojure`
```

That's about it - please let me know how it works for you, and I'll add it all to the README.

Cheers - V

**Comment by anonimitoraf** [comment]
Thanks @vemv - I tried it out and successfully started my program as usual. However when I use `cider-connect` to connect via SSH, I don't see javadocs (like what you posted [here](https://clojurians.slack.com/archives/C0617A8PQ/p1699381762708279))

- I've got `cider-enrich-classpath` set to `t`
- CIDER 1.12.0
- `nrepl/1.1.0`
- `cider-nrepl/0.43.1`

**Comment by vemv** [comment]
Thanks!

For discarding possibilities, it would be useful to try a minimal local project first?

**Comment by anonimitoraf** [comment]
Sure, I've created https://github.com/anonimitoraf/barebones-clojure and did the same steps as in your recording (I think)


https://github.com/clojure-emacs/enrich-classpath/assets/15933322/ac7df83e-d741-4513-99db-c621374a4529

**Comment by vemv** [comment]
Nice, thank you!

Could you share your JVM process classpath?

As found with `ps aux | grep java`

**Comment by anonimitoraf** [comment]
No worries

```
ubuntu     77325  8.0 26.1 3145468 494556 pts/0  Sl+  10:30   1:00 /usr/bin/java -XX:-OmitStackTraceInFastThrow -XX:-OmitStackTraceInFastThrow -Dclojure.main.report=stderr --add-opens=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED --add-opens=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED -Dclojure.basis=.cpcache/676205888.basis -classpath src:resources:/home/ubuntu/.m2/repository/org/clojure/clojure/1.11.1/clojure-1.11.1.jar:/home/ubuntu/.m2/repository/org/clojure/core.specs.alpha/0.2.62/core.specs.alpha-0.2.62.jar:/home/ubuntu/.m2/repository/org/clojure/spec.alpha/0.3.218/spec.alpha-0.3.218.jar:/home/ubuntu/.m2/repository/cider/cider-nrepl/0.43.1/cider-nrepl-0.43.1.jar:/home/ubuntu/.m2/repository/nrepl/nrepl/1.1.0/nrepl-1.1.0.jar:/home/ubuntu/.m2/repository/cider/orchard/0.19.0/orchard-0.19.0.jar:/home/ubuntu/.m2/repository/mx/cider/logjam/0.1.1/logjam-0.1.1.jar:/home/ubuntu/.cache/mx.cider/enrich-classpath/17081/815684688/1900108823.jar clojure.main -m com.anonimito.barebones-clojure
```

**Comment by vemv** [comment]
Got it, you need to install the JDK sources in Ubuntu

https://docs.cider.mx/cider/troubleshooting.html#navigation-to-jdk-sources-doesnt-work

(as you can see those instructions are for JDK17; if you're using something newer, I'd appreciate a docs update)

**Comment by vemv** [comment]
...At this point, given that Enrich wraps both Lein and tools.deps, we could proactively ensure that those sources are there e.g. fail-fast when not found.

(Users can always disable Enrich if that requirement cannot be immediately satisfied)

**Comment by anonimitoraf** [comment]
Great, installing `openjdk-17-source openjdk-17-doc` worked! Thanks so much

**Comment by vemv** [comment]
Cheers, enjoy!

Any kind of feedback always welcome. There's a small refinement coming https://github.com/clojure-emacs/cider/issues/3576

**Comment by anonimitoraf** [comment]
Oh yea, regarding docs, how can I do the same for third-party libs? For example: https://stripe.dev/stripe-java ?

**Comment by vemv** [comment]
Yes, if you depend on a given Maven library, they always include (per the Maven rules) a "sources" jar.

enrich-classpath fetches those.

**Comment by anonimitoraf** [comment]
I see. I can see that Stripe, for example indeed do have docs and sources:

```
~/.m2/repository/com/stripe/stripe-java/22.12.0$ ls -al
total 28236
drwxr-xr-x 2 ubuntu ubuntu     4096 Nov 10 07:13 .
drwxr-xr-x 3 ubuntu ubuntu     4096 Nov  7 12:39 ..
-rw-rw-r-- 1 ubuntu ubuntu      288 Nov 10 07:13 _remote.repositories
-rw-rw-r-- 1 ubuntu ubuntu 19695356 Nov 10 07:13 stripe-java-22.12.0-javadoc.jar
-rw-rw-r-- 1 ubuntu ubuntu       40 Nov 10 07:13 stripe-java-22.12.0-javadoc.jar.sha1
-rw-rw-r-- 1 ubuntu ubuntu  1775886 Nov 10 07:13 stripe-java-22.12.0-sources.jar
-rw-rw-r-- 1 ubuntu ubuntu       40 Nov 10 07:13 stripe-java-22.12.0-sources.jar.sha1
-rw-r--r-- 1 ubuntu ubuntu  7402879 Nov  7 12:42 stripe-java-22.12.0.jar
-rw-r--r-- 1 ubuntu ubuntu       40 Nov  7 12:42 stripe-java-22.12.0.jar.sha1
-rw-r--r-- 1 ubuntu ubuntu     1767 Nov  7 12:39 stripe-java-22.12.0.pom
-rw-r--r-- 1 ubuntu ubuntu       40 Nov  7 12:39 stripe-java-22.12.0.pom.sha1
```

but I'm not seeing docs in CIDER inspector 🤔

**Comment by vemv** [comment]
Will try soon!

Sometimes it happens that there's some issue in the .java files that makes them unparseable ([example](https://github.com/aws/aws-sdk-java/issues/3045))

**Comment by anonimitoraf** [comment]
I see, no worries/rush and thanks.

Unrelated, I've noticed that when using the `clojure.sh` script, my CPU usage is constantly at 100% as opposed to ~1% when using just vanilla Clojure CLI

For context, I'm using `systemd`:

`ExecStart=/bin/bash -c './clojure.sh clojure -M:prod'`

<img width="1440" alt="Screenshot 2023-11-11 at 8 45 13 am" src="https://github.com/clojure-emacs/enrich-classpath/assets/15933322/1498c02b-95bb-46d8-baa4-dbe37e500092">

vs

`ExecStart=/bin/bash -c 'clojure -M:prod'`

<img width="1426" alt="Screenshot 2023-11-11 at 8 42 57 am" src="https://github.com/clojure-emacs/enrich-classpath/assets/15933322/79e4c1d5-b5d6-4a50-80d1-2446c056d957">

(If it helps, here's the rest of my `systemd` service config)
```
[Unit]
Description=API service
After=network.target
StartLimitIntervalSec=0
# Always restart service
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=1
User=ubuntu

[Install]
WantedBy=multi-user.target
```

**Comment by anonimitoraf** [comment]
Actually, my bad, I should've checked if this is the case without `systemd`. Checking this now

**Comment by vemv** [comment]
Interesting, hadn't experienced it (I would have noticed since I use the infamous Intel MBP 🛫)

There are a few ways in which you can determine what your CPU is doing

* The one I know best is opening YourKit (there's a free trial), going to the Threads tab, searching for threads with a 🔥 icon, and clicking on one for checking its stacktrace.

Example:

<img width="879" alt="image" src="https://github.com/clojure-emacs/enrich-classpath/assets/1162994/aec290e0-72e4-4bb1-9d5d-82920024a961">

* The other one is, within the terminal repl started by clojure.sh hitting `Ctrl + \` so that the JVM will dump stacktraces from every single thread

Example:

<img width="764" alt="image" src="https://github.com/clojure-emacs/enrich-classpath/assets/1162994/b9d1c5e3-f63a-4549-b70b-e512f37666e8">

You could share the entire output with me over here or privately.

* Lastly, there are other profilers that may easily pinpoint the busiest threads.

**Comment by vemv** [comment]
The simplest explanation is that it's performing [this](https://github.com/clojure-emacs/cider-nrepl/blob/5b62ae3d478fd17d1c455758841c7e65475f59af/src/cider/nrepl.clj#L70-L72) one-off job. On my machine (which as mentioned can be very noisy) it never hogs the CPU and it completes reasonably fast.

It also completes quickly on cider-nrepl's CI matrix.

Are you sure that the 100% last more than a few minutes?

**Comment by vemv** [comment]
Regarding Stripe, the underlying parsing errors come from the fact that is uses Lombok at build time, but it's not a Maven dependency. So without it in the classpath, the Java parser will complain.

They even had some historical issues related to Lombok. It's a plugin which has a mixed fame. https://github.com/stripe/stripe-java/issues?q=is%3Aissue+lombok

Note that, however, jump-to-definition works.

**Comment by anonimitoraf** [comment]
Cool thanks for the info. I'll send through peofiling results tonight. From my observation, CPU usage hangs around at 100% for a few minutes after the program has finished startup, I'll check later if this goes on for longer

Re: Stripe - I see, no worries!

**Comment by vemv** [comment]
Cheers. https://github.com/clojure-emacs/orchard/issues/211 would solve all issues (Stripe parsing and CPU hogging - assuming we've guessed the cause correctly).

It might become a reality before the year wraps up!
